File: libs/sdk-dotnet/AuthzSdk/Authorization/PermissionMatcher.cs
namespace AuthzSdk.Authorization;

/// <summary>
/// Matches required permissions against user permissions, including wildcard support.
/// Semantically equivalent to the Java PermissionMatcher.
/// </summary>
public static class PermissionMatcher
{
    /// <summary>
    /// Checks whether the user permission set satisfies the required permission.
    /// Supports exact match and wildcard patterns (e.g. "estoque.*").
    /// </summary>
    public static bool Matches(HashSet<string> userPermissions, string requiredPermission)
    {
        ArgumentNullException.ThrowIfNull(userPermissions);
        ArgumentException.ThrowIfNullOrWhiteSpace(requiredPermission);

        if (userPermissions.Contains(requiredPermission))
        {
            return true;
        }

        if (requiredPermission.EndsWith(".*", StringComparison.Ordinal))
        {
            var prefix = requiredPermission[..^2];
            var prefixWithDot = prefix + ".";

            return userPermissions.Any(permission => permission.StartsWith(prefixWithDot, StringComparison.Ordinal));
        }

        foreach (var permission in userPermissions)
        {
            if (!permission.EndsWith(".*", StringComparison.Ordinal))
            {
                continue;
            }

            var prefix = permission[..^2];
            if (requiredPermission.StartsWith(prefix + ".", StringComparison.Ordinal))
            {
                return true;
            }
        }

        return false;
    }
}

File: libs/sdk-java/src/main/java/com/platform/authz/sdk/aop/PermissionMatcher.java
package com.platform.authz.sdk.aop;

import java.util.Objects;
import java.util.Set;

/**
 * Matches required permissions against the user permission set, including wildcard support.
 */
public class PermissionMatcher {

    public boolean matches(Set<String> userPermissions, String requiredPermission) {
        Objects.requireNonNull(userPermissions, "userPermissions must not be null");
        Objects.requireNonNull(requiredPermission, "requiredPermission must not be null");

        if (userPermissions.contains(requiredPermission)) {
            return true;
        }

        if (requiredPermission.endsWith(".*")) {
            String prefix = requiredPermission.substring(0, requiredPermission.length() - 2);
            return userPermissions.stream()
                    .anyMatch(permission -> permission.startsWith(prefix + "."));
        }

        return userPermissions.stream()
                .filter(permission -> permission.endsWith(".*"))
                .map(permission -> permission.substring(0, permission.length() - 2))
                .anyMatch(prefix -> requiredPermission.startsWith(prefix + "."));
    }
}
